### ISSUE TITLE
Is there a blockUntil implementation?

### ISSUE THREAD

**Issue by chaviw** [ReactiveX/RxJava#4336]
I was searching through the docs and there doesn't seem to be any `blockUntil`. There's a `waitUntil` and `doUntil`, but I was hoping for something a bit different. 

It would be nice to have a way to hold data until something triggers the "unblock". This way, after the "unblock" each piece of data can be grabbed without losing anything.

Are there plans to build a `blockUntil`? I'd be happy to help contribute on that.

**Comment by akarnokd** [comment]
I can't remember any `waitUntil` or `doUntil` in RxJava; are you sure you posted to the right library? There was [an operator PR](https://github.com/ReactiveX/RxJava/pull/3781) `pressureValve` that offered a second channel to pause a main channel losslessly and non-blockingly, but it's on hold for now.

**Comment by chaviw** [comment]
My bad, I meant `skipUntil` and `takeUntil`, but none of those block the emission

**Comment by chaviw** [comment]
And yes, I think that pr is similar to what I want

**Comment by JakeWharton** [comment]
That sounds like it would be named `bufferWhile`

**Comment by chaviw** [comment]
Sure, that makes sense. So I'm guessing there's nothing in RxJava that handles that yet. If no ones working on it, then I can take a shot at it. Basically it would queue up data (maybe up to a specified limit) and then begins emissions after the condition is met

**Comment by akarnokd** [comment]
What is your use case?

**Comment by chaviw** [comment]
To be as specific without going into implementation details:
As soon as the app is opened, "A" is started. The execution is asynchronous. The user can still interact with the app by invoking "B". "B" should NEVER start until "A" is done otherwise things will break. We want to queue up all the "B" tasks so after "A" is done, all the "B" interaction will get handled

An additional note, we can't stop the user from invoking "B" before "A" is done. Not because of requirements, but because of the nature of mobile devices. 

In our first attempt, we just canceled "A" when "B" occurred and started again. This can lead to a slower start time for the user. 

The current implementation waits for "A" to finish and then grabs the last state "B" to invoke. That's obviously not ideal because we can possibly lose data

**Comment by akarnokd** [comment]
Sounds like you need `concatEager` that prestarts all sources but keeps ordering among them.

**Comment by chaviw** [comment]
That might work. I'll try that out and see if it works for our use-case. Thanks

**Comment by chaviw** [comment]
I dug a bit more into `concatEager` and it's not exactly what I need. 

I think I found another way to fix my issue and not lose data. However, I still think having a `bufferWhile` would be nice. It would look something like `bufferWhile(Observable<U> other)` which will begin emitting once the `other` Observable emits an item

**Comment by akarnokd** [comment]
I'm closing this issue due to inactivity. If you have further input on the issue, don't hesitate to reopen this issue or post a new one.
